### PR TITLE
server: create api to trigger table metadata update job

### LINKED ISSUE #128897: Create new HTTP API to triggering the table metadata update job
In order to provide customers will more fine grained control over the table metadata update job, a new HTTP api will be created to manually trigger the table metadata update job. This API will support a query parameter, `onlyIfStale` which will only trigger the job if the system table is deemed stale (and not already running).

Jira issue: CRDB-41265

Epic CRDB-37558

### PR DESCRIPTION
**Pull request by kyle-a-wong**
adds a new API v2 endpoint,
`POST /api/v2/table_metadata/updatejob`.

server: create api to trigger table metadata update job

adds a new API v2 endpoint,
`POST /api/v2/table_metadata/updatejob`.

This new API will trigger the table metadata update job
to run and refresh the system.table_metadata table.

This will only trigger the job if it isn't already
running. If the `?onlyIfStale` query param is
provided, the job won't be triggered unless deemed
stale, determined by the following setting:
 `obs.tablemetadata.data_valid_duration`

note that `onlyIfStale` with any value other then
`false` will be treated as true.

Resolves: https://github.com/cockroachdb/cockroach/issues/128897
Epic: [CRDB-37558](https://cockroachlabs.atlassian.net/browse/CRDB-37558)
Release note: None

### CODE REVIEW

File: pkg/server/api_v2_databases_metadata_test.go [hunk 1]
```diff
@@ -512,29 +519,161 @@ func TestGetTableMetadataUpdateJobStatus(t *testing.T) {
...
+WHERE id = $1 AND claim_instance_id IS NOT NULL`, jobs.UpdateTableMetadataCacheJobID)
+		require.NoError(t, err)
+		if !row.Next() {
+			return errors.New("no node has claimed the job")
+		}
+		require.NoError(t, row.Scan(&nodeID))
+
+		rpcGatewayNode := (nodeID + 1) % 3
+		_, err = testCluster.Server(rpcGatewayNode).GetStatusClient(t).UpdateTableMetadataCache(ctx,
+			&serverpb.UpdateTableMetadataCacheRequest{Local: false})
```
Comment (xinhaoz) [inline pkg/server/api_v2_databases_metadata_test.go:571]: I think we need to delete this part which is triggering via status server.

File: pkg/server/api_v2_databases_metadata.go [hunk 2]
```diff
@@ -752,6 +763,31 @@ func (a *apiV2Server) getTableMetadataUpdateJobStatus(
...
+// triggerTableMetadataUpdateJob will trigger the table metadata update job if it isn't currently running and if it
+// is stale, if onlyIfStale is true.
+func (a *apiV2Server) triggerTableMetadataUpdateJob(
+	ctx context.Context, onlyIfStale bool,
+) (tmJobTriggeredResponse, error) {
+	status, err := a.getTableMetadataUpdateJobStatus(ctx)
+	if err != nil {
+		return tmJobTriggeredResponse{}, err
+	}
+	if status.CurrentStatus != "NOT_RUNNING" {
```
[the remark this answers is not in the document]
Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata.go]: Done

File: pkg/server/api_v2_databases_metadata_test.go (hunk 1 above)
Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata_test.go]: Removed this and switched to using a test knob to indicate when the job is ready

File: pkg/sql/tablemetadatacache/update_table_metadata_cache_job.go [hunk 3]
```diff
@@ -38,13 +38,19 @@ var _ jobs.Resumer = (*tableMetadataUpdateJobResumer)(nil)
...
 func (j *tableMetadataUpdateJobResumer) Resume(ctx context.Context, execCtxI interface{}) error {
-	log.Infof(ctx, "starting table metadata update job")
 	j.job.MarkIdle(true)
 
 	execCtx := execCtxI.(sql.JobExecContext)
 	metrics := execCtx.ExecCfg().JobRegistry.MetricsStruct().
 		JobSpecificMetrics[jobspb.TypeUpdateTableMetadataCache].(TableMetadataUpdateJobMetrics)
-
+	var onJobStartKnob func()
+	var onJobCompleteKnob func()
```
Comment (xinhaoz) [inline pkg/sql/tablemetadatacache/update_table_metadata_cache_job.go:47]: If you want you can combine this into one line
var onJobStartKnob, onJobCompleteKnob, onJobReady func()

File: pkg/server/api_v2_databases_metadata_test.go [hunk 4]
```diff
@@ -544,29 +548,122 @@ func TestGetTableMetadataUpdateJobStatus(t *testing.T) {
...
 }
 
-func makeApiRequest[T any](t *testing.T, client http.Client, uri string) (mdResp T) {
-	req, err := http.NewRequest("GET", uri, nil)
+func TestTriggerMetadataUpdateJob(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	jobCompletedChan := make(chan interface{})
+	jobReadyChan := make(chan interface{})
+	defer close(jobCompletedChan)
```
Comment (xinhaoz) [inline pkg/server/api_v2_databases_metadata_test.go:577]: Do we need to close the other channel too?

File: pkg/server/api_v2_databases_metadata_test.go [hunk 5]
```diff
@@ -544,29 +548,122 @@ func TestGetTableMetadataUpdateJobStatus(t *testing.T) {
...
+	skip.UnderStress(t, "too slow under stress")
```
Comment (xinhaoz) [inline pkg/server/api_v2_databases_metadata_test.go:578]: Let's put this skip before the channel definitions.

File: pkg/server/api_v2_databases_metadata_test.go [hunk 6]
```diff
@@ -544,29 +548,122 @@ func TestGetTableMetadataUpdateJobStatus(t *testing.T) {
...
+					OnJobComplete: func() {
+						jobCompletedChan <- struct{}{}
+					},
+				},
+			},
+		},
+	})
+	ctx := context.Background()
+	defer testCluster.Stopper().Stop(ctx)
+	conn := testCluster.ServerConn(0)
```
Comment (xinhaoz) [inline pkg/server/api_v2_databases_metadata_test.go:596]: If the queries executed in this connection should always succeed, you can wrap this in `sqlutils.MakeSQLRunner` so you don't have to check the error every time.

File: pkg/server/status.go [hunk 7]
```diff
@@ -4211,7 +4211,7 @@ func (s *statusServer) localUpdateTableMetadataCache() (
 	select {
 	case s.updateTableMetadataJobSignal <- struct{}{}:
 	default:
-		return nil, status.Errorf(codes.Unavailable, "update table metadata cache job is not ready to start execution")
+		return nil, status.Errorf(codes.Aborted, "update table metadata cache job is not ready to start execution")
```
Comment (xinhaoz) [inline pkg/server/status.go:4214]: Just curious, why the switch to aborted here?

**Review by xinhaoz — APPROVED** [review]
Nice! Excited to connect this to the new pages. Mostly testing comments - feel free to merge after addressing them.

File: pkg/sql/tablemetadatacache/update_table_metadata_cache_job.go (hunk 3 above)
Comment (kyle-a-wong) [reply pkg/sql/tablemetadatacache/update_table_metadata_cache_job.go]: Done

File: pkg/server/api_v2_databases_metadata_test.go (hunk 6 above)
Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata_test.go:598]: Done

File: pkg/server/api_v2_databases_metadata_test.go (hunk 5 above)
Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata_test.go]: Done

File: pkg/server/status.go (hunk 7 above)
Comment (kyle-a-wong) [reply pkg/server/status.go:4214]: I think that typically, the unavailable status should be used for when the service is temporarily unavailable, where as the aborted status should be used when an operation is aborted. To me this case falls under an latter.

File: pkg/server/api_v2_databases_metadata_test.go (hunk 4 above)
Comment (kyle-a-wong) [reply pkg/server/api_v2_databases_metadata_test.go:579]: Done

**Comment by kyle-a-wong** [comment]
tftr

bors r+

**Pull request merged**